### ISSUE TITLE
Problem with exclamation marks in YML

### ISSUE THREAD

**Issue by jszroberto** [geofffranks/spruce#75]
Hi everyone, I was trying to use spruce  in the following way: 
```
bosh download manifest cf.yml 
spruce merge cf.yml mytemplate.yml 
```

And I found that spruce apparently has issues with ! , and the manifest deployed by Bosh contains them: 
```
properties: 
    route_registrar:
      routes:
      - name: uaa
        port: 8080
        tags:
          component: uaa
        uris:
        - uaa.fabric.ng.bluemix.net
        - ! '*.uaa.fabric.ng.bluemix.net'
```
The error that I get is: 
```
tmp/cf.yml: unmarshal []byte to yaml failed: yaml: line 525: did not find URI escaped octet

```

Any idea of what I could do? Thanks in advance

**Comment by filefrog** [comment]
I believe you are running into YAML tag syntax here.  See http://stackoverflow.com/questions/9664113/what-does-a-single-exclamation-mark-do-in-yaml

Try quoting the string value in its entirety:

```
properties: 
    route_registrar:
      routes:
      - name: uaa
        port: 8080
        tags:
          component: uaa
        uris:
        - uaa.fabric.ng.bluemix.net
        - "! '*.uaa.fabric.ng.bluemix.net'"
```

**Comment by jszroberto** [comment]
Actually, the strange is that the manifest upload doesn't have those, it's bosh who is adding them: 

route_registrar:      
      routes:     
      - name: uaa     
        port: 8080
        tags:     
          component: uaa    
        uris:
        - uaa.fabric.lon.bluemix.net      
        - '*.uaa.fabric.lon.bluemix.net' 

Single quotes should be the same, right?

**Comment by filefrog** [comment]
That *is* strange.  I've never done much with downloaded BOSH manifests; virtually all of my use cases involve keeping template files in a git repo, sprucing them together, and deploying that.

Have you asked the BOSH team about this?

**Comment by jszroberto** [comment]
Issue in Bosh created, but do you have any idea why spruce is not able to handle exclamations?

**Comment by filefrog** [comment]
The _unspecified tag_ is for signalling to the YAML processor that what follows is a custom object type (i.e. not a string or numeric builtin).  For example, if you want to encode a Foo as a YAML, *directly*, you would do something like this:

```yml
myfoo: !Foo <foo internal representation>
```

I don't know if BOSH is trying to trigger this behavior or not, but the YAML parser is trying to oblige and is unable to.  A general purpose utility like Spruce can't be written to "just know" what other tools are going to produce, so it has to stick to the built-in types.

We'll see what the BOSH team says.

**Comment by geofffranks** [comment]
Beginning a value with `!` has a specific meaning in YAML (signifies local data types, used for serialization). If you need to begin a value with an exclamation mark, simply quote it.

**Comment by jszroberto** [comment]
@geofffranks  it wasn't me who added those exclamation marks, there were added by Bosh when "bosh download manifest". Bosh team  says that it's generated by the standard and they cannot do anything about it. I will try to skip the problem with a regexp to before calling spruce. Thanks for the help :+1:

**Comment by HeavyWombat** [comment]
I ran into the same issue. It turns out this seems to be caused by the underlying YAML library. I as able to trace it to [this issue in go-yaml](https://github.com/go-yaml/yaml/issues/75). There is already a [merge request with a potential fix](https://github.com/go-yaml/yaml/pull/133), but it is not processed yet.

So I tried it myself. I have two valid YAMLs, which cannot be spruce-merged due to this issue. I patched my local go-yaml library based on the fix and recompiled it as well as spruce itself. After doing that, I was able to spruce-merge the files.

As this is a problem in an underlying library, there is nothing spruce can do about it (other than using a modified forked version of the go-yaml lib). I just wanted to put it here in case somebody runs into the same issue.
